sympy/core/basic.py

```diff
@@ -33,6 +33,63 @@ def as_Basic(expr):
             expr))
 
 
+# This must be a base class of Basic so that __init_subclass__ applies to Basic
```

Comment:
Let's see if we can just put `__init_subclass__` on `Basic`, and move any logic that should also apply to `Basic` itself to `Basic.__init__`.

Reply:
`__init_subclass__` modifies subclass when it is created by `type` (i.e. where the line `class mysubclass(Basic):` is)

i.e.:
```python
class A:
    def __init_subclass__(cls):
        print('foo')

class B(A):
    pass
```

Will print `'foo'` eventhough you never instantiate `B`. I don't think `__init__` can take over what `__init_subclass__` is doing here.

Reply:
We can refactor the code from `Basic.__init_subclass__` to a separate function `_prepare_class_assumptions` and then do:
```python
class Basic(Printable):
    def __init_subclass__(cls):
        _prepare_class_assumptions(cls)

# __init_subclass__ is not called on Basic itself
_prepare_class_assumptions(Basic)
```

Reply:
Does `Basic` actually need these modifications itself? Or in other words, is there a reason to use a non-subclassed `Basic` object?

Reply:
It's used in the Basic tests https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_basic.py.

Reply:
I think you could also just do `Basic.__init_subclass__(Basic)`.

Reply:
I've refactored this as described.

Reply:
looks good to me